server/src/game/world.ts

```diff
@@ -140,69 +131,56 @@ export class World {
       for (const unit of units) {
         world.addUnit(unit);
       }
-      world.updateCivTileVisibility(Number(civID));
-    }
-
-    for (const leaderID in leaderPool) {
-      const leaderData = data.leaderPool[leaderID];
-      world.leaderPool[Number(leaderID)] = Leader.import(leaderData);
-      if (leaderData.civID !== null) {
-        world.setCivLeader(leaderData.civID, Number(leaderID));
+      if (civData.leader !== null) {
+        world.updateLeaderTileVisibility(leaders[civData.leader]);
+        world.setDomainLeader(civData.domainID, leaders[civData.leader]);
       }
     }
 
     for (const cb of callbacks) {
-      cb(world);
+      cb(world, leaders);
     }
 
     return world;
   }
 
-  getUpdates(): { (civID: number): Event }[] {
+  getUpdates(): { (leader: Leader): Event }[] {
     // TODO: more updates?
     return this.map.getUpdates().concat(this.updates.splice(0));
   }
 
-  // leaders
-  getLeaderPool(): [LeaderData[], LeaderData[]] {
-    const leaderList: LeaderData[] = [];
-    const takenLeaderList: LeaderData[] = [];
-
-    for (const id in this.leaderPool) {
-      const leader = this.leaderPool[id];
-      if (leader.isTaken()) {
-        takenLeaderList.push(leader.getData());
-      } else {
-        leaderList.push(leader.getData());
-      }
-    }
-
-    return [leaderList, takenLeaderList];
-  }
-
-  // leaders, civs
-  setCivLeader(civID: number, leaderID: number): boolean {
-    const leader = this.leaderPool[leaderID];
-    if (!leader || leader.isTaken()) {
+  /**
+   * Sets the leader of a domain if none already exists. *Cannot* transfer a domain from one leader to another.
+   * @param domainID 
+   * @param leader 
+   * @returns 
+   */
+  setDomainLeader(domainID: DomainID, leader: Leader): boolean {
+    const domain = this.getDomain(domainID);
+    if (!domain || domain.hasLeader()) {
       return false;
     }
 
-    if (this.civs[civID].leader) {
-      this.civs[civID].leader?.unselect();
-    }
-    this.civs[civID].leader = leader;
-    leader.select(civID);
-    for (const unit of this.civs[civID].getUnits()) {
-      unit.knowledge = {};
-      unit.updateKnowledge(leader.startingKnowledge);
-    }
+    domain.setLeader(leader);
+    leader.addDomain(domain);
+
+    this.updates.push((leader) => ['leaderUpdate', [leader.id, leader.getData()]]);
 
     return true;
   }
 
-  // civs
-  getCiv(civID: number): Civilization {
-    return this.civs[civID];
+  getDomain(domainID: DomainID): Domain {
+    if (isCivDomain(domainID)) return this.getCiv(domainID);
+    else if (isCityDomain(domainID)) return this.getCity(domainID);
+    else throw new ValueError('Bad domain ID type: ' + domainID.type);
```

Comment:
Is this possible? Won't it be a compile error instead of a potential runtime error if you remove this check? (I'm not sure, but in some languages, it certainly would be a compile error.)

You might have to do something else than just removing the check... but there should be _some_ way to do this safely with compile time checks.

Reply:
Just removing the check won't work, no. I agree with you in principle, though, so if you can think of a way to do it, I definitely think we should.